### PR TITLE
Add options to customize parsing/stringifying JSON

### LINKED ISSUE #1274: parseJson option
As discussed there: https://github.com/sindresorhus/ky/issues/254#issuecomment-629613861

Option should allow to pass a user-function that will perform JSON parsing.

Use cases:

1. Use `bourne` to mitigate the hyped prototype pollution.
2. Parse JSON with `reviver`

Example API:

```typescript

type ParseJsonFunction = <OUTPUT = unknown>(text: string) => OUTPUT

const myGot = got.extend({
  parseJson: text => bourne(text),
})
```

Should be available both via `.extend()` and as an option to each request.

### PR DESCRIPTION
**Pull request by Giotino**
Custom function to parse JSON responses.

#### Checklist

- [x] I have read the documentation.
- [x] I have included a pull request description of my changes.
- [x] I have included some tests.
- [x] If it's a new feature, I have included documentation updates.

Fixes #1274

### CODE REVIEW

File: source/core/index.ts [hunk 1]
```diff
@@ -224,6 +227,7 @@ export interface Defaults {
 	allowGetBody: boolean;
 	https?: HTTPSOptions;
 	methodRewriting: boolean;
+	parseJson?: ParseJsonFn;
```
Comment (szmarczak) [inline source/core/index.ts:230]:
```suggestion
	parseJson: ParseJsonFn;
```

Comment (szmarczak) [inline source/core/index.ts:230]:
```suggestion
	parseJson: ParseJsonFn;
```

**Comment by szmarczak** [comment]
@Giotino Can you also do `stringifyJson`?

**Comment by szmarczak** [comment]
IMO this is good to merge.

**Comment by szmarczak** [comment]
Actually I'm having second thoughts. This also adds more complexity. What's against having

```js
const {body} = await got(...);
const parsed = JSON.parse(body, parser);
```

**Comment by Giotino** [comment]
I thought so, but with `parseJson` you can have an extended got that always behave the same without having to explicitly use your parser every time.

It's not a "must have"  feature, but since someone showed interest in that it means that it might be nice to have. Also I don't think it adds that much complexity.

**Comment by Giotino** [comment]
Maybe we should ear more opinions before proceeding.

File: readme.md [hunk 2]
```diff
@@ -317,6 +317,23 @@ const body = await got(url).json();
...
 ```
 
+###### parseJson
+
+Type: `(text: string) => unknown`\
+Default: `(text) => JSON.parse(text)`
+
+Function used to parse JSON responses.
+
+Example:
```
Comment (sindresorhus) [inline readme.md:327]: It's not immediately clear to the user what this example does different.

See https://github.com/sindresorhus/ky/pull/257/files#diff-0730bb7c2e8f9ea2438b52e419dd86c9R349-R367 for better docs.

File: readme.md [hunk 3]
```diff
@@ -317,6 +317,23 @@ const body = await got(url).json();
 const body = await got(url, {responseType: 'json', resolveBodyOnly: true});
 ```
 
+###### parseJson
+
+Type: `(text: string) => unknown`\
+Default: `(text) => JSON.parse(text)`
```
Comment (sindresorhus) [inline readme.md:323]:
```suggestion
Default: `(text: string) => JSON.parse(text)`
```

File: readme.md [hunk 4]
```diff
@@ -317,6 +317,23 @@ const body = await got(url).json();
...
+
+Function used to parse JSON responses.
+
+Example:
+
+```js
```
Comment (sindresorhus) [inline readme.md:329]:
```suggestion
```js
const got = require('got');
```

File: source/core/index.ts [hunk 5]
```diff
@@ -118,6 +118,7 @@ type CacheableRequestFn = (
 ) => CacheableRequest.Emitter;
 
 type CheckServerIdentityFn = (hostname: string, certificate: DetailedPeerCertificate) => Error | void;
+export type ParseJsonFn = (text: string) => unknown;
```
Comment (sindresorhus) [inline source/core/index.ts:121]:
```suggestion
export type ParseJsonFunction = (text: string) => unknown;
```

I know we haven't been consistent with this, but never too late to start.

Comment (Giotino) [reply source/core/index.ts:121]: Ok, I'm also going to change the other `*Fn` types, is it ok?

**Comment by kirillgroshkov** [comment]
> Actually I'm having second thoughts. This also adds more complexity. What's against having
> 
> ```js
> const {body} = await got(...);
> const parsed = JSON.parse(body, parser);
> ```

In our case we have ~50 methods that call API via `got(...).json()`. To add a reviver to them right now means adding it to 50 places and risking that we forget to add it to newly added methods. Having it built in the extended got means the change in only 1 place and safety to add more methods.

Same thinking as why got.extend function and hooks exist in the first place. Without it - it'll need copy/pasting the hooks to every usage.

**Comment by szmarczak** [comment]
Hmm... @kirillgroshkov is right. So `stringifyJson` is also welcome, but unnecessary. It's up to you :)

**Comment by Giotino** [comment]
Examples needs to be fixed/improved.

@sindresorhus @szmarczak
While working on the documentation I noted that the examples are incosistent.
Examples of examples :)
```js
const got = require('got');

await got();
```

```js
const got = require('got');

(async () => {
  await got();
})();
```
```js
await got();
```

I think that `const got = require('got');` and the async wrapper are not needed, I can work on a PR normalizing all the examples.

**Comment by szmarczak** [comment]
> I think that const got = require('got'); and the async wrapper are not needed

@sindresorhus has said a while ago that these examples should be 100% functional. Copy-paste and it should work.

File: source/core/index.ts [hunk 6]
```diff
@@ -120,6 +120,7 @@ type CacheableRequestFunction = (
 
 type CheckServerIdentityFunction = (hostname: string, certificate: DetailedPeerCertificate) => Error | void;
 export type ParseJsonFunction = (text: string) => unknown;
+export type StringifyJsonFunction = (object: any) => string;
```
Comment (sindresorhus) [inline source/core/index.ts:123]:
```suggestion
export type StringifyJsonFunction = (object: unknown) => string;
```

Makes no practical difference, but we prefer `unknown` whenever possible. Same with the other places `any` is used.

File: readme.md [hunk 7]
```diff
@@ -1,4 +1,4 @@
-<div align="center">
+Œ<div align="center">
```
Comment (sindresorhus) [inline readme.md]: I have no idea what happened in this commit. Looks like the GitHub web UI completely borked up the commit.

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/got/pull/1298/files#r433074774 still needs to be resolved.

**Comment by sindresorhus** [comment]
> I think that const got = require('got'); and the async wrapper are not needed, I can work on a PR normalizing all the examples.

That would be great as a separate PR. We should normalize to **always** require Got and always use the async IIFE.

**Comment by Giotino** [comment]
> https://github.com/sindresorhus/got/pull/1298/files#r433074774 still needs to be resolved.

Yeah, I'm aware of it, I was kinda busy so I delayed it a bit.

File: readme.md (hunk 2 above)
Comment (Giotino) [reply readme.md:327]: I can't figure out any meaningful example for `stringifyJson`, currently I added a random one that transforms all numbers in strings.

File: readme.md [hunk 8]
```diff
@@ -326,13 +326,18 @@ Function used to parse JSON responses.
 
 Example:
 
+Using Bourne to prevent prototype pollution
```
Comment (szmarczak) [inline readme.md:329]:
```suggestion
- Using Bourne to prevent prototype pollution:

```

File: readme.md [hunk 9]
```diff
@@ -344,15 +349,22 @@ Function used to stringify JSON requests body.
 
 Example:
 
+All numbers as strings
```
Comment (szmarczak) [inline readme.md:352]:
```suggestion
- All numbers as strings:

```

File: readme.md (hunk 2 above)
Comment (szmarczak) [reply readme.md:327]: A better example would be something that skips keys that begin with `_`.

Comment (sindresorhus) [reply readme.md:327]: > A better example would be something that skips keys that begin with _.

Yes, in addition ot the current one with numbers.

**Comment by sindresorhus** [comment]
Excellent 👌

**Pull request merged**